application.cpp

```diff
@@ -449,10 +466,12 @@ void application::exec() {
          // execute the highest priority item
          more = pri_queue.execute_highest();
       }
+      pri_queue.clear(); // if io_serv->stop() was called, queue may not be empty and hold io_serv pointers
```

Comment:
Instead of `pri_queue.clear();`, seems like we should `while( pri_queue.execute_highest() ) {};`

Reply:
Yes, I was not sure. There is actually a `execute_all()` api which might be even better, but I was not sure if it was appropriate to execute more actions after quit() was called. Sounds like it is.

Reply:
Individual posted lambdas can test `is_quiting()` if it matters to them. We already drain all the queue in the while loop after `quit()` except for possibly the last one.

Reply:
done, thank you @heifner

Reply:
That doesn't feel right to me at all. I would expect asio::io_context-like behavior where queued but not dispatched handlers are simply discarded when the io_context is stopped and then destroyed. For example,
```c++
int main() {
	boost::asio::io_context ctx;

	boost::asio::defer(ctx, []() { std::cout << "a" << std::endl;});
	boost::asio::defer(ctx, []() { std::cout << "b" << std::endl;});
	boost::asio::defer(ctx, [&](){ctx.stop();});
	boost::asio::defer(ctx, []() { std::cout << "c" << std::endl;});
	boost::asio::defer(ctx, []() { std::cout << "d" << std::endl;});

	ctx.run();
	return 0;
}
```
This will not print `c` nor `d` even though they were queued before stopping (yes technically this contrived simple example is UB in this area due to non-guaranteed ordering).

When interacting with objects that are wrapping asio, I think they really need to adhere to asio-like behaviors.

Reply:
The existing behavior is to drain the queue (except for corner case of potentially the last item in the queue).

Argument for draining: Briefly the idea that something important might be posted to the queue that needs to be done; some kind of cleanup or shutdown action. Examples: sync file to disk, report failure to external systems (gelf log), flush chainbase to disk, etc.

Argument for not draining: We post blocks to be processed to the main thread. If something goes wrong and we `quit()` the application, it seems best to stop right away. Continuing to attempt to process queued up blocks could make matters worse; maybe corrupting SHiP log or chainbase state for example.

Seems we should have sync and error reporting or other highly important must do operations on shutdown in plugin `shutdown()` and not in `post()`ed tasks. I believe that is the case today. 

Gelf logging already uses io_context directly so it will be stopped immediately. Also since gelf logging is using app io_context directly that means it is the highest priority item that runs. That is clearly wrong and should be fixed. https://github.com/AntelopeIO/leap/issues/650

I'm in favor of changing the behavior to not drain the queue, but to stop immediately mimicking `asio::io_context`.

Reply:
OK, I believe this should do it @heifner . Thanks for the input as well @spoonincode .